tools/src/com.oracle.truffle.tools.thermometer/src/com/oracle/truffle/tools/thermometer/Thermometer.java

```diff
@@ -0,0 +1,217 @@
+import com.oracle.truffle.api.instrumentation.StandardTags;
+import com.oracle.truffle.api.source.Source;
+import sun.misc.Signal;
+
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Level;
+
+public final class Thermometer implements AutoCloseable {
+
+    private final Env env;
+
+    private final AtomicLong loadedSource = new AtomicLong();
+    private final ThermometerSampler sampler = new ThermometerSampler();
+    private final AtomicReference<ThermometerState> previousState = new AtomicReference<>();
+
+    private boolean running;
+    private long startTime;
+    private long lastObservedCompilationActivity;
+    private long lastObservedCompilationTime;
+    private int interruptAfterStable;
+    private long maxHeapSize;
+    private PrintStream logStream;
+    private ScheduledThreadPoolExecutor executor;
+
+    private boolean hasIterationLocation;
+
+    public Thermometer(Env env) {
+        this.env = env;
+        lastObservedCompilationActivity = 0;
+        maxHeapSize = 0;
+    }
+
+    public synchronized void start(ThermometerConfig config) {
+        if (running) {
+            throw new IllegalStateException("the thermometer is already running");
+        }
+
+        running = true;
+        startTime = System.nanoTime();
+        lastObservedCompilationTime = startTime;
+        interruptAfterStable = config.getInterruptAfterStable();
+
+        // Install a node at the root of each compilation unit to set the compilation flag
+
+        final SourceSectionFilter rootNodeFilter = SourceSectionFilter.newBuilder().tagIs(StandardTags.RootTag.class).build(); // FIXME this is supposed to be CompilationRootTag!
+        env.getInstrumenter().attachExecutionEventFactory(rootNodeFilter, context -> new ThermometerCompilationSampleNode(sampler));
+
+        // Install a node at the IPS sample location to count iterations
+
+        hasIterationLocation = config.getIterationLocation() != null;
+
+        if (hasIterationLocation) {
+            final int index = config.getIterationLocation().lastIndexOf(':');
+
+            if (index == -1) {
+                env.getLogger("").log(Level.WARNING, "thermometer iteration location could not be parsed - should be file:line");
+                hasIterationLocation = false;
+            } else {
+                final String file = config.getIterationLocation().substring(0, index);
+                final int line = Integer.parseInt(config.getIterationLocation().substring(index + 1));
+
+                final SourceSectionFilter iterationNodeFilter = SourceSectionFilter.newBuilder()
+                        .tagIs(StandardTags.RootTag.class) // Thermometer TODO: I want the root of compilation units, so I catch OSR methods
```

Comment:
I want a tag for compilation roots. So including OSR roots. Do we think we can do that?

Reply:
Instrumentation does not really know about this, and probably shouldn't.
How about you instrument a StatementTag instead?
Then you could use CompilerDirectives.isCompilationRoot() that folds to a constant on the fast path?

Reply:
`StatementTag` would trigger in so many more places, and have a lot more footprint overhead, it sounds suboptimal to me.

Reply:
It does work with a statement tag, yes (it doesn't work with the root tag unless you turn off OSR.) As Benoit says it adds quite a bit of overhead.

I don't know how to argue whether instrumentation 'should' know about it or not, apart from this is an instrument it does need to know about it to solve the problem optimally. Also, an event that's triggered once per compilation unit entry seems pretty general to me.

Reply:
I was assuming that the iteration is something that specified as a single line location so very little instrumentation nodes would get created no?

Reply:
Sorry there's some confusion here. There's two probe nodes.

One probes iterations per second, and there's only one instance in the program (multiple due to splitting.)

The other sets a flag to say whether we're in compiled code or not, which is then sampled every few ms to give a proportion of time spent in compiled code. This second probe is the one we're ideally like to have once per compilation unit.

Reply:
Ok. So no problem with counting the iteration then.

Detecting OSR is tricky. We could maybe do it like with a new API CompilerDirectives.getCompilationRoot():

```
if (CompilerDirectives.inCompilationRoot()) {
  boolean isMethod = instrumentedNode.getRootNode().getCallTarget() == CompilerDirectives.getCompilationRoot();
  CompilerAsserts.partialEvaluationConstant(isMethod);
  if (isMethod) {
    // not in OSR
  } else {
    // in OSR, but hard to know which one, probably needs another API.
    CompilerDirectives.getCompilationRoot().getRootNode().somehowGetToTheLoop();
  }
}
```

I suggest we skip OSR in the first version.

I think you have another problem here. You actually need to track whether in compiled code for each thread. You could use ContextThreadLocals for that purpose. It is not very efficient in multi-threaded mode atm, but I have plans to make it much more efficient.

Another idea is to leverage the CPUSampler#takeSample() which can provide you with the same info, but is currently more heavy on overhead as it collects more info.